### ISSUE TITLE
"ConstantInputWarning: An input array is constant; correlation coefficient is not defined"

### ISSUE THREAD

**Issue by higorsilvaa** [UKPLab/sentence-transformers#2729]
I'm fine-tuning my model (neuralmind/bert-large-portuguese-cased) for STS task using assin dataset, which contains pairs in Portuguese for RTE and STS tasks, I'm facing a problem, when I try 1 epoch, it says "ConstantInputWarning: An input array is constant; correlation coefficient is not defined", which causes my Pearson and Spearman metrics to result in NaN. But, when I try for 2 epochs, the fine-tuning works perfectly.
Also, if I fine-tune the model with 1 epoch but remove the last 2 pairs from the training set, it also works.
I can't explain what might be happening. Is it a bug in the framework?

**Comment by tomaarsen** [comment]
Hello!

Huh, that is quite odd. I think that warning would only happen when calling the `EmbeddingSimilarityEvaluator` and either the model predictions of the similarity are all identical, or the gold evaluation similarities are all identical. Why this would happen with 1 epoch but not for 2: I'm not sure. My guess is that somehow perhaps the `scores` that you gave to `EmbeddingSimilarityEvaluator` are all identical? Is that possible?

I'm experimenting with roughly your setup here:
```python
from datasets import load_dataset
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
    SentenceTransformerModelCardData,
)
from sentence_transformers.losses import CoSENTLoss
from sentence_transformers.similarity_functions import SimilarityFunction
from sentence_transformers.training_args import BatchSamplers
from sentence_transformers.evaluation import EmbeddingSimilarityEvaluator, TripletEvaluator

import logging

logging.basicConfig(level=logging.INFO)

# 1. Load a model to finetune with 2. (Optional) model card data
model = SentenceTransformer(
    "neuralmind/bert-large-portuguese-cased",
    model_card_data=SentenceTransformerModelCardData(
        language="pt",
        license="mit",
    )
)

# 3. Load a dataset to finetune on
dataset = load_dataset("nilc-nlp/assin", "full")
# Convert the dataset to 2 text columns and a score column: https://sbert.net/docs/sentence_transformer/training_overview.html#dataset-format
train_dataset = dataset["train"].select_columns(["premise", "hypothesis", "relatedness_score"]).rename_column("relatedness_score", "score")
eval_dataset = dataset["validation"].select_columns(["premise", "hypothesis", "relatedness_score"]).rename_column("relatedness_score", "score")
test_dataset = dataset["test"].select_columns(["premise", "hypothesis", "relatedness_score"]).rename_column("relatedness_score", "score")

# 4. Define a loss function
loss = CoSENTLoss(model)

# 5. (Optional) Specify training arguments
args = SentenceTransformerTrainingArguments(
    # Required parameter:
    output_dir="models/bert-large-portuguese-assin",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=32,
    learning_rate=2e-5,
    warmup_ratio=0.1,
    fp16=False,  # Set to False if you get an error that your GPU can't run on FP16
    bf16=True,  # Set to True if you have a GPU that supports BF16
    # Optional tracking/debugging parameters:
    eval_strategy="steps",
    eval_steps=20,
    save_strategy="steps",
    save_steps=20,
    save_total_limit=2,
    logging_steps=20,
    run_name="bert-large-portuguese-assin",  # Will be used in W&B if `wandb` is installed
)

# 6. (Optional) Create an evaluator & evaluate the base model
dev_evaluator = EmbeddingSimilarityEvaluator(
    sentences1=eval_dataset["premise"],
    sentences2=eval_dataset["hypothesis"],
    scores=eval_dataset["score"],
    main_similarity=SimilarityFunction.COSINE,
    name="assin-dev",
)
dev_evaluator(model)

# 7. Create a trainer & train
trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss=loss,
    evaluator=dev_evaluator,
)
trainer.train()

# (Optional) Evaluate the trained model on the test set
test_evaluator = EmbeddingSimilarityEvaluator(
    sentences1=test_dataset["premise"],
    sentences2=test_dataset["hypothesis"],
    scores=test_dataset["score"],
    main_similarity=SimilarityFunction.COSINE,
    name="assin-test",
)
test_evaluator(model)

# 8. Save the trained model
model.save_pretrained("models/bert-large-portuguese-assin/final")

# 9. (Optional) Push it to the Hugging Face Hub
# model.push_to_hub("bert-large-portuguese-assin")
```
This seems to give pretty normal results for me (e.g. 64.21 Spearman Correlation on assin dev before training -> abut 80 on dev after training).

- Tom Aarsen

**Comment by higorsilvaa** [comment]
Thank you for the rapid response.

**Comment by higorsilvaa** [comment]
Thanks for your quick reply.
The difference between our codes is the loss function. I was using CosineSimilarityLoss, as well as the training_stsbenchmark.py example.
So I tried using CoSENTLoss, just like you, and it worked.
I didn't understand. Do you have any idea what might be happening?
Thanks again for helping me.

**Comment by higorsilvaa** [comment]
Thanks for your quick reply.
The difference between our codes is the loss function. I was using CosineSimilarityLoss, as well as the training_stsbenchmark.py example.
So I tried using CoSENTLoss, just like you, and it worked for 1 epoch, but no for 2.
I tried to specify the lr  and enable the bf16 like you and aparently it worked.
